numba/core/withcontexts.py

```diff
@@ -464,3 +465,64 @@ def _mutate_with_block_callee(blocks, blk_start, blk_end, inputs, outputs):
         block=ir.Block(scope=scope, loc=loc),
         outputs=outputs,
     )
+
+class _ParallelChunksize(WithContext):
+    is_callable = True
+
+    """A context-manager that remembers and saves away the current
+    chunksize for executing parfors and then changes that chunksize
+    to the programmer specified value.  At the end of the region,
+    the original chunksize is restored.
+    """
+    def mutate_with_body(self, func_ir, blocks, blk_start, blk_end,
+                         body_blocks, dispatcher_factory, extra):
+        ir_utils.dprint_func_ir(func_ir, "Before with changes", blocks=blocks)
+        assert extra is not None
+        args = extra["args"]
+        assert len(args) == 1
+        arg = args[0]
+        scope = blocks[blk_start].scope
+        loc = blocks[blk_start].loc
+        if isinstance(arg, ir.Arg):
+            arg = ir.Var(scope, arg.name, loc)
+
+        set_state = []
+        restore_state = []
+
+        # global for Numba itself
+        gvar = ir.Var(scope, ir_utils.mk_unique_var("$ngvar"), loc)
+        set_state.append(ir.Assign(ir.Global('numba', numba, loc), gvar, loc))
+        # getattr for set chunksize function in Numba
+        spcattr = ir.Expr.getattr(gvar, 'set_parallel_chunksize', loc)
+        spcvar = ir.Var(scope, ir_utils.mk_unique_var("$spc"), loc)
+        set_state.append(ir.Assign(spcattr, spcvar, loc))
+        # call set_parallel_chunksize
+        orig_pc_var = ir.Var(scope, ir_utils.mk_unique_var("$save_pc"), loc)
+        cs_var = ir.Var(scope, ir_utils.mk_unique_var("$cs_var"), loc)
+        set_state.append(ir.Assign(arg, cs_var, loc))
+        spc_call = ir.Expr.call(spcvar, [cs_var], (), loc)
+        set_state.append(ir.Assign(spc_call, orig_pc_var, loc))
+
+        restore_spc_call = ir.Expr.call(spcvar, [orig_pc_var], (), loc)
+        restore_state.append(ir.Assign(restore_spc_call, orig_pc_var, loc))
```

Comment:
I think loc at restore_state as referring to start block location is correct.  It would seem even weirder if you did the loc after the block which may be some unrelated line of code.